engine/src/flutter/impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -50,496 +52,581 @@ class PositionWriter {
+        AppendVertices(origin, perpendicular);
+      }
+      AddCap(cap, origin, perpendicular, false);
+    } else if (with_close) {
+      // Closed contour, join back to origin
+      FML_DCHECK(origin == origin_point_);
+      FML_DCHECK(last_point_ == origin);
+      AddJoin(join_, origin, last_perpendicular_, origin_perpendicular_);
+
+      last_perpendicular_ = origin_perpendicular_;
+      last_point_ = origin;
     } else {
-      direction = (polyline.GetPoint(point_i) - polyline.GetPoint(point_i - 1))
-                      .Normalize();
+      AddCap(cap_, last_point_, last_perpendicular_.GetVector(), false);
     }
-    return SeparatedVector2(Vector2{-direction.y, direction.x},
-                            stroke_width * 0.5f);
+    has_prior_segment_ = false;
   }
 
-  void AddVerticesForLinearComponent(PositionWriter& vtx_builder,
-                                     const size_t component_start_index,
-                                     const size_t component_end_index,
-                                     const size_t contour_start_point_i,
-                                     const size_t contour_end_point_i,
-                                     const Path::PolylineContour& contour) {
-    bool is_last_component = component_start_index ==
-                             contour.components.back().component_start_index;
-
-    for (size_t point_i = component_start_index; point_i < component_end_index;
-         point_i++) {
-      bool is_end_of_component = point_i == component_end_index - 1;
-
-      Point offset_vector = offset.GetVector();
-
-      vtx.position = polyline.GetPoint(point_i) + offset_vector;
-      vtx_builder.AppendVertex(vtx.position);
-      vtx.position = polyline.GetPoint(point_i) - offset_vector;
-      vtx_builder.AppendVertex(vtx.position);
-
-      // For line components, two additional points need to be appended
-      // prior to appending a join connecting the next component.
-      vtx.position = polyline.GetPoint(point_i + 1) + offset_vector;
-      vtx_builder.AppendVertex(vtx.position);
-      vtx.position = polyline.GetPoint(point_i + 1) - offset_vector;
-      vtx_builder.AppendVertex(vtx.position);
-
-      previous_offset = offset;
-      offset = ComputeOffset(point_i + 2, contour_start_point_i,
-                             contour_end_point_i, contour);
-      if (!is_last_component && is_end_of_component) {
-        // Generate join from the current line to the next line.
-        join_proc(vtx_builder, polyline.GetPoint(point_i + 1),
-                  previous_offset.GetVector(), offset.GetVector(),
-                  scaled_miter_limit, scale);
-      }
+ private:
+  PositionWriter& vtx_builder_;
+  const Scalar half_stroke_width_;
+  const Scalar maximum_join_cosine_;
+  const Scalar minimum_miter_cosine_;
+  const Join join_;
+  const Cap cap_;
+  const Scalar scale_;
+  const Tessellator::Trigs& trigs_;
```

Comment:
Raw pointer, same thing here.

Reply:
Trigs is tricky because it is not copyable because they are shared/cached and hold a vector. I would have to move their vector into a shared pointer in order to pass them around by value. Take a look at the class in the Tessellator class and recommend something.

Reply:
I left this aspect of the implementation for now as fixing `Trigs` is going to affect a lot of places.

Reply:
https://github.com/flutter/flutter/issues/167758